feat(discovery): port kubeApi discovery to v3

tthvo opened the pull request:
# Welcome to Cryostat3! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat3/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] [Signed all commits using a GPG signature](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#gpg-commit-signature-verification)

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Fixes: #238 

## Description of the change:

Port KubeDiscoveryAPI from v2 to v3.

## Motivation for the change:

See #238

## How to manually test:

Build the application and push to quay.io:

```bash
IMAGE_NAMESPACE=quay.io/thvo
./mvnw clean package -DskipTests=true && \
podman tag quay.io/cryostat/cryostat:3.0.0-snapshot ${IMAGE_NAMESPACE}/cryostat:3.0.0-snapshot && \
podman push ${IMAGE_NAMESPACE}/cryostat:3.0.0-snapshot
```

Install the application with Helm Chart (https://github.com/cryostatio/cryostat-helm/pull/128):

```bash
helm install cryostat3 ./charts/cryostat/ \
  --set core.discovery.kubernetes.enabled=true \
  --set core.discovery.kubernetes.namespaces="{default,myns,.}" \
  --set core.discovery.kubernetes.portNames="{jmx-jfr,my-port}" \
  --set core.discovery.kubernetes.portNumbers="{9091,9097}" \
  --set core.image.repository=${IMAGE_NAMESPACE}/cryostat
```

Deploy/Scale/Teardown sample applications:

```bash
cd cryostat-operator && make sample_app # or make undeploy_sample_app
```

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,540 @@
+                                                                new EndpointsHandler(),
+                                                                ENDPOINTS_INFORMER_RESYNC_PERIOD));
+                                        logger.info(
+                                                String.format(
```
`logger.info(String.format())` should be replaced with `logger.infov()` :-)

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,540 @@
+                .collect(Collectors.toList());
+    }
+
+    class KubeConfig {
```
`@ApplicationScoped static`, maybe?

tthvo commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,540 @@
+import org.jboss.logging.Logger;
+
+@ApplicationScoped
+class OpenShiftDiscovery extends KubeApiDiscovery {
```
Ahh I see! Thanks for the explanation! I will remove that!

andrewazores commented:
In 2.4, the NAMESPACES env var was expected to be simply a comma-separated list of namespaces that Cryostat should try to set up an informer for. If the env var was not set then it would default to an empty list - no discovery.

I think in 3.0 we want to preserve that same behaviour, however with one addition: a special character that signifies "your own namespace", as in the namespace that the Cryostat Pod is running within. I suggest `.` for this special character. This can be included in the namespace list, so it would be possible for the user to set the variable to `a,b,.`. Cryostat should try to look up what its own namespace is, map `.` to that value, then deduplicate the namespace list (in case `.` is equal to `a` or `b`). This way configuration is a little bit simpler. I think we can also default the namespace list to the `.` singleton rather than defaulting to empty.

tthvo commented:
> In 2.4, the NAMESPACES env var was expected to be simply a comma-separated list of namespaces that Cryostat should try to set up an informer for. If the env var was not set then it would default to an empty list - no discovery.

> I think in 3.0 we want to preserve that same behaviour, however with one addition: a special character that signifies "your own namespace", as in the namespace that the Cryostat Pod is running within. I suggest . for this special character. This can be included in the namespace list, so it would be possible for the user to set the variable to a,b,.. Cryostat should try to look up what its own namespace is, map . to that value, then deduplicate the namespace list (in case . is equal to a or b). This way configuration is a little bit simpler. I think we can also default the namespace list to the . singleton rather than defaulting to empty.

Oh thank you! This behaviour with `.` makes sense to me. I suppose we would want the default namespace list to empty and require it to be explicitly set  in a similar way to other configurations (e.g. `cryostat.discovery.k8s.port-names`)?

andrewazores commented:
Sure, I think for the individual Cryostat container level it makes sense to default to empty lists there. We can apply default values in the Helm chart, in Operator CRs, and maybe also in Operator default webhooks.

tthvo commented:
Sorry for the long delay! It took quite a lot of time to learn and try different approaches :D 

In this case, the discovery plugin responds Endpoint event by fetching all current endpoints and doing a diff with persisted targets. Also, I added a `parent` field that references the parent node (similar to `ownerReferences` in k8s) to trace up the tree and prune nodes.

It seems to work just fine? Not sure if I chose the right approach. Please let me know what you think :D

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+            };
+
+    @Transactional
+    void onStart(@Observes @Priority(1) StartupEvent evt) {
```
What is this Priority for?

tthvo replied:
Oh this is to ensure the Realm node is persisted first before initializing the informers. During my testing, I see that the informers call the event handler right away when created so if it's called in the same transaction as the realm node creation, the realm is not yet available.

tthvo replied:
maybe there is a nicer way around it?

andrewazores replied:
Okay, makes sense. Just add a comment here explaining why it has the Priority then please.

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+
+    @Transactional
+    void onStart(@Observes @Priority(1) StartupEvent evt) {
+        if (!(enabled())) {
```
Minor style nitpick: `!(enabled())` / `!isAvailable()` inconsistent parentheses

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+    boolean enabled;
+
+    @ConfigProperty(name = "cryostat.discovery.k8s.port-names")
+    Optional<List<String>> JmxPortNames;
```
`jmxPortNames`, `jmxPortNumbers` (lowercase)

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+    @ConfigProperty(name = "cryostat.discovery.kubernetes.enabled")
+    boolean enabled;
+
+    @ConfigProperty(name = "cryostat.discovery.k8s.port-names")
```
`cryostat.discovery.k8s` and `cryostat.discovery.kubernetes` - I think using the full `kubernetes` is preferable

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+            DiscoveryNode node =
+                    DiscoveryNode.getNode(
+                                    n -> {
+                                        return name.equals(n.name)
```
So this label is used for the case where Namespace A and Namespace B both have Endpoints objects named E, is that right? This way we can unambiguously use the `<name, namespace>` pair as an identifier for that particular node.

tthvo replied:
Oh yes it is indeed for identify nodes of the same names in different namespaces. I used it mostly for identifying intermediate parent nodes tho (i.e. pod, replicaset, deployment, etc.).

> This way we can unambiguously use the <name, namespace> pair as an identifier for that particular node.

That means we should return `Pair<HasMetadata, <Pair<String, DiscoveryNode>>` from this method `queryForNode`?

andrewazores replied:
Oh no, you don't need to make any changes here, I was just trying to make sure my reading of it was correct.

tthvo replied:
Tho, that being said, that should also check for nodeType in case there might be 2 nodes of the same name, in the same namespace but different type.

andrewazores commented on src/main/resources/application-test.properties:
```diff
@@ -3,6 +3,11 @@ quarkus.smallrye-openapi.info-title=Cryostat API (test)
 cryostat.discovery.jdp.enabled=true
 cryostat.discovery.podman.enabled=true
 cryostat.discovery.docker.enabled=true
+cryostat.discovery.kubernetes.enabled=true
```
I think you can remove these lines from the `-test` and `-dev` profiles - these are applied as overrides on top of the default `application.properties`, so if the desired values are the same then they can be left off so they default to the standard profile.

andrewazores commented on src/main/java/io/cryostat/discovery/NodeType.java:
```diff
@@ -0,0 +1,143 @@
+    }
+}
+
+enum KubeDiscoveryNodeType implements NodeType {
```
Would it make sense to leave `BaseNodeType` in here, but move `KubeDiscoveryNodeType` into `KubeApiDiscovery` and `ContainerDiscoveryNodeType` into `ContainerDiscovery`?

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+                URI connectUrl = URI.create(jmxUrl.toString());
+
+                try {
+                    return Target.getTargetByConnectUrl(connectUrl);
```
I think it would be extremely uncommon, but this could yield an unexpected result if there is a URL collision between KubeApiDiscovery and another discovery mechanism, and the other mechanism has created a Target with this URL first. It seems like it would become possible for the Endpoints event handler to end up with this query getting executed and pulling a Target from a different Realm out of the database and returning it.

So I think the Target either needs to be found by traversing this Realm's subtree, or else it can be queried directly from the database like this but then the Realm subtree membership needs to be verified before returning it.

tthvo replied:
Right! I think that makes sense! I can verify its associated node if its of type `Endpoint`.

andrewazores replied:
That works too. If it happens then I would say log a warning and return null, I guess? Or maybe just throw an `IllegalStateException` since this should be a very rare occurrence and only if there is a URL collision like that, which really means that there are overlapping Realms. Until https://github.com/cryostatio/cryostat3/issues/71 is solved, overlapping Realms should be considered an invalid configuration.

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+    @ApplicationScoped
+    static final class KubeConfig {
+        public static final String KUBERNETES_NAMESPACE_PATH =
+                "/var/run/secrets/kubernetes.io/serviceaccount/namespace";
```
I think this could also be a `@ConfigProperty`, but this path should be used as the default value in `application.properties`.

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+    private final class EndpointsHandler implements ResourceEventHandler<Endpoints> {
+        @Override
+        public void onAdd(Endpoints endpoints) {
+            logger.infov(
```
#349 - I think the log level for these kinds of events should probably be `debug`

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+public class KubeApiDiscovery {
+    public static final String REALM = "KubernetesApi";
+
+    public static final long ENDPOINTS_INFORMER_RESYNC_PERIOD = Duration.ofSeconds(30).toMillis();
```
This should also be a `@ConfigProperty`, something like `cryostat.discovery.kubernetes.resync-period=30s`

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+    Optional<List<String>> JmxPortNames;
+
+    @ConfigProperty(name = "cryostat.discovery.k8s.port-numbers")
+    Optional<List<Integer>> JmxPortNumbers;
```
What happens if you just declare this as `List<Integer>` and have `cryostat.discovery.kubernetes.port-numbers=` in `application.properties`? IIRC the config framework will give you an empty list already in this case. I think the current behaviour you've implemented is already correct, so I just wonder if the `Optional<>` can be cleaned up a little this way.

tthvo replied:
I guess `Optional` is necessary then. If I remove it, there is an error at startup:

```
Failed to load config value of type java.util.List<java.lang.String> for: cryostat.discovery.kubernetes.namespaces
```

tthvo replied:
Yeah it seems so...it works fine with non-empty values tho. Just when it's empty, there is an error above :((

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+    boolean available() {
+        try {
+            boolean hasNamespace = StringUtils.isNotBlank(kubeConfig.getOwnNamespace());
+            return kubeConfig.kubeApiAvailable() || hasNamespace;
```
Should this be `&&`?

tthvo replied:
I actually ported this from `2.x`. Though, it might be beneficial for environment such as local `kubeConfig` where the fabric can pick up info from there? Otherwise, I think `&&` sounds good.

andrewazores replied:
The local `kubeConfig` idea is interesting, but I guess there would be other things to implement to support it, and I don't think anything like that is planned right now.

tthvo replied:
Sounds good I will update it to `&&` for now :D

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,623 @@
+            return serviceHost.isPresent();
+        }
+
+        KubernetesClient kubeClient() {
```
Does this need to be lazy-initialized like this? It's already only happening inside of the lazily initialized `KubeConfig`, so I think it would be fine to just initialize the `kubeClient` directly where it's declared.

Minor side bonus to doing it that way and making it `final` is that it's also threadsafe.

andrewazores reviewed (commented):
Thanks for doing all this, Thuan! All my review comments are just style or refactor/cleanup really, but the actual substance and structure of this PR looks great.

tthvo commented:
> Thanks for doing all this, Thuan! All my review comments are just style or refactor/cleanup really, but the actual substance and structure of this PR looks great.

I am happy to help out anytime ^^ Updated and addressed all reviews now :D

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,702 @@
+        return informers;
+    }
+
+    @Transactional
```
I don't think this annotation will work as intended, since this method is not an "entrypoint" that the Quarkus framework calls into but rather one that is invoked from other application code.

https://quarkus.io/guides/transaction#programmatic-approach

I think this example is probably suitable:

```java
        QuarkusTransaction.joiningExisting().run(() -> {
            //do work
        });
```

andrewazores replied:
No, I think that one is also just buggy :-)

tthvo replied:
Ahh okayy haha! I will update that!

andrewazores commented on src/main/java/io/cryostat/discovery/KubeApiDiscovery.java:
```diff
@@ -0,0 +1,702 @@
+        }
+    }
+
+    private void pruneOwnerChain(DiscoveryNode nsNode, Target target) {
```
I think this also needs to operate within a transaction, since it calls `persist()` and `delete()`

tthvo commented:
Fixed now :D @andrewazores is the lastest commit what you are looking for?

andrewazores commented:
Yes, I think that looks good now. I'll run it through some testing soon.

andrewazores commented:
/build_test

andrewazores commented:
Brief manual testing looks good, so I think this is good to go after the test failure is corrected.

tthvo commented:
/build_test

tthvo commented:
/build_test

tthvo commented:
Phew tests failure and spotbugs errors are fixed now^^ I did manual tests with container, jdp and custom discovery as this PR also affects them. They are all fine.
